pandas/tests/io/test_gcs.py

```diff
@@ -70,6 +70,67 @@ def mock_get_filepath_or_buffer(*args, **kwargs):
     tm.assert_frame_equal(df1, df2)
 
 
+@td.skip_if_no("gcsfs")
+@pytest.mark.parametrize("encoding", ["utf-8", "utf-16"])
+def test_to_csv_compression_gcs(monkeypatch, compression_only, encoding):
+    """
+    Compression and encoding should work if binary mode is requested.
+
+    "gs://" is internally converted to a file object, to_csv supports compression for
+    file objects in bianry mode but not in text mode.
+
+    GH 35677
+    """
+    from fsspec import AbstractFileSystem, registry
+
+    registry.target.clear()  # noqa  # remove state
+    df1 = DataFrame(
+        {
+            "int": [1, 3],
+            "float": [2.0, np.nan],
+            "str": ["t", "s"],
+            "dt": date_range("2018-06-18", periods=2),
+        }
+    )
+
+    # reference of compressed and encoded file
+    compression = {"method": compression_only}
+    if compression_only == "gzip":
+        compression["mtime"] = 1  # be reproducible
+    buffer = BytesIO()
+    df1.to_csv(buffer, compression=compression, encoding=encoding, mode="wb")
+
+    s = BytesIO()
+    s.close = lambda: True
+    s.flush = lambda: None
+
+    class MockGCSFileSystem(AbstractFileSystem):
+        def open(*args, **kwargs):
+            s.seek(0)
+            return s
+
+    monkeypatch.setattr("gcsfs.GCSFileSystem", MockGCSFileSystem)
+
+    # write compressed file with explicit compression
+    df1.to_csv(
+        "gs://test/test.csv", compression=compression, encoding=encoding, mode="wb",
+    )
+    assert s.getvalue() == buffer.getvalue()
+
+    # write compressed file with implicit compression
+    s.seek(0)
+    if compression_only == "gzip":
+        compression_only = "gz"
+    compression["method"] = "infer"
+    df1.to_csv(
+        "gs://test/test.csv." + compression_only,
+        compression=compression,
+        encoding=encoding,
+        mode="wb",
```

Comment:
Can you test also with the default `mode='w'`? In order to avoid `b` to be forced to make it work.

Reply:
We have to have `mode="wb"` internally to support compression for file objects. The question is whether pandas internally sets "wb"  or whether the user has to set it. I think it is best to let a core member make a decision on that (not me).

Reply:
my opinion: if it doesn't break anything else, I would convert `mode` for all file names that are internally converted to file objects to binary mode. A good place to do that would probably be `get_filepath_or_buffer`. I would also like to move `compression = infer_compression(path_or_buf, compression)` inside `get_filepath_or_buffer` to prevent possible future errors if `get_filepath_or_buffer` is called before `infer_compression`.